File: apps/api/app/schemas/learning.py
from pydantic import BaseModel, Field


class LearningRequest(BaseModel):
    topic: str = Field(
        min_length=1,
        max_length=200,
    )
    language: str = Field(
        default="English",
        min_length=1,
        max_length=50,
    )
    level: str = Field(
        default="beginner",
        min_length=1,
        max_length=50,
    )


class ExampleItem(BaseModel):
    title: str
    explanation: str
    code: str | None = None


class QuizQuestion(BaseModel):
    question: str
    options: list[str]
    answer: str
    explanation: str


class PracticeItem(BaseModel):
    question: str
    hint: str


class Assignment(BaseModel):
    title: str
    description: str
    requirements: list[str]


class MiniProject(BaseModel):
    title: str
    description: str
    requirements: list[str]


class TopicSearchCard(BaseModel):
    title: str
    slug: str
    category: str
    description: str


class TopicSearchResponse(BaseModel):
    query: str
    results: list[TopicSearchCard]


class LearningResponse(BaseModel):
    topic: str
    # 1. WhyLearnThis
    why_learn_this: str
    # 2. ProfessionalDefinition
    professional_definition: str
    # 3. EasyExplanation
    easy_explanation: str
    # 4. RealWorldAnalogy
    real_world_analogy: str
    # 5. Applications
    real_world_applications: list[str]
    # 6. Syntax (optional/core concepts)
    syntax_or_core_concepts: str
    # 7. Examples
    examples: list[ExampleItem]
    # 8. CommonMistakes
    common_mistakes: list[str]
    # 9. Practice
    interactive_practice: list[PracticeItem]
    # 10. Quiz (5 MCQs)
    quiz: list[QuizQuestion]
    # 11. Assignment
    assignment: Assignment
    # 12. MiniProject
    mini_project: MiniProject
    # 13. RelatedTopics
    related_topics: list[str]
    # 14. Summary
    summary: str = ""
    # Next topic suggestion
    next_topic: str = ""
    # Cache indicator
    cached: bool = False

File: apps/web/lib/api/learning.ts
import { apiRequest } from './client';
import { getToken } from './auth';

export type LearningRequest = {
  topic: string;
  language?: string;
  level?: string;
};

export type ExampleItem = {
  title: string;
  explanation: string;
  code: string | null;
};

export type QuizQuestion = {
  question: string;
  options: string[];
  answer: string;
  explanation: string;
};

export type PracticeItem = {
  question: string;
  hint: string;
};

export type Assignment = {
  title: string;
  description: string;
  requirements: string[];
};

export type MiniProject = {
  title: string;
  description: string;
  requirements: string[];
};

export type TopicSearchCard = {
  title: string;
  slug: string;
  category: string;
  description: string;
};

export type TopicSearchResponse = {
  query: string;
  results: TopicSearchCard[];
};

export type LearningResponse = {
  topic: string;
  why_learn_this: string;
  professional_definition: string;
  easy_explanation: string;
  real_world_analogy: string;
  real_world_applications: string[];
  syntax_or_core_concepts: string;
  examples: ExampleItem[];
  common_mistakes: string[];
  interactive_practice: PracticeItem[];
  quiz: QuizQuestion[];
  assignment: Assignment;
  mini_project: MiniProject;
  related_topics: string[];
  summary?: string;
  next_topic: string;
  cached?: boolean;
};

export async function searchKnowledgeTopics(
  query: string,
  token?: string,
): Promise<TopicSearchResponse> {
  const authToken = token ?? getToken();
  const params = new URLSearchParams();
  if (query) params.append('q', query);

  return apiRequest<TopicSearchResponse>(`/learning/search?${params.toString()}`, {
    method: 'GET',
    token: authToken || undefined,
  });
}

export async function generateLearningContent(
  data: LearningRequest,
  token?: string,
): Promise<LearningResponse> {
  const authToken = token ?? getToken();

  return apiRequest<LearningResponse>('/learning/generate', {
    method: 'POST',
    token: authToken || undefined,
    body: {
      topic: data.topic,
      language: data.language ?? 'English',
      level: data.level ?? 'beginner',
    },
  });
}
